File: ray-tracer-fs/Shapes.fs
module Shapes

open Vectors.Vec3F64

type Ray =
  { origin: Vec3
    direction: Vec3 }

  member r.at(t: float) = r.origin + t * r.direction

let inline ray origin direction =
  { origin = origin
    direction = direction }

type Hit =
  { t: float
    p: Vec3
    normal: Vec3
    frontFace: bool }

type Hittable =
  abstract member rayHit: Ray -> minT: float -> maxT: float -> Hit option

type Sphere =
  { center: Vec3
    radius: float }

  member inline s.rayHit ray minT maxT = (s :> Hittable).rayHit ray minT maxT

  interface Hittable with
    member s.rayHit ray minT maxT =

      let oc = ray.origin - s.center
      let halfB = oc.dot ray.direction
      let c = oc.lengthSquared () - s.radius * s.radius
      let discriminant = halfB * halfB - c

      if discriminant < 0. then
        None
      else
        let sqrtD = sqrt discriminant
        let mutable t = (-halfB - sqrtD)

        if t < minT || t > maxT then
          t <- (-halfB + sqrtD)

          if t < minT || t > maxT then
            t <- -1

        if t < 0 then
          None
        else
          let p = ray.at t
          let normal = (p - s.center) / s.radius
          let frontFace = ray.direction.dot normal < 0.

          if not frontFace then
            normal *= -1.

          Some(
            { t = t
              p = p
              normal = normal
              frontFace = frontFace }
          )


let inline sphere center radius = { center = center; radius = radius }


type HittableList(hs: Hittable seq) =
  member inline s.rayHit ray minT maxT = (s :> Hittable).rayHit ray minT maxT

  interface Hittable with
    member _.rayHit ray minT maxT =
      let mutable closestSoFar = maxT
      let mutable closestHit = None

      for h in hs do
        match h.rayHit ray minT closestSoFar with
        | Some hit ->
          closestSoFar <- hit.t
          closestHit <- Some hit
        | None -> ()

      closestHit

File: web/ray-tracer-ts/shapes.ts
import { Vec3 } from "./vec3";
import { Ray, type Hit, type Hittable } from "./ray";

export class Sphere implements Hittable {
  constructor(
    public center: Vec3,
    public radius: number,
  ) {}

  rayHit(ray: Ray, minT: number, maxT: number): Hit | null {
    const oc = ray.origin.sub(this.center);
    const halfB = oc.dot(ray.direction);
    const c = oc.lengthSquared() - this.radius * this.radius;
    const discriminant = halfB * halfB - c;

    if (discriminant < 0) {
      return null;
    }

    const sqrtD = Math.sqrt(discriminant);
    let t = -halfB - sqrtD;

    // Check if t is in valid range
    if (t < minT || t > maxT) {
      t = -halfB + sqrtD;
      if (t < minT || t > maxT) {
        return null;
      }
    }

    const p = ray.at(t);
    let normal = p.sub(this.center).div(this.radius);
    const frontFace = ray.direction.dot(normal) < 0;

    // Flip normal if hitting back face
    if (!frontFace) {
      normal = normal.mul(-1);
    }

    return { t, p, normal, frontFace };
  }
}

export class HittableList implements Hittable {
  constructor(public objects: Hittable[]) {}

  rayHit(ray: Ray, minT: number, maxT: number): Hit | null {
    let closestSoFar = maxT;
    let closestHit: Hit | null = null;

    for (const obj of this.objects) {
      const hit = obj.rayHit(ray, minT, closestSoFar);
      if (hit) {
        closestSoFar = hit.t;
        closestHit = hit;
      }
    }

    return closestHit;
  }
}
